AK/BigIntBase.h

```diff
@@ -0,0 +1,566 @@
+
+// ===== Static storage for big integers =====
+template<size_t total_size>
+struct StackBuffer {
+    NativeWord data[total_size];
+    NativeWord* free_ptr = data;
+
+    constexpr NativeWord* allocate(size_t size)
+    {
+        NativeWord* result = free_ptr;
+        free_ptr += size;
+        VERIFY(free_ptr <= data + total_size);
+        return result;
+    }
+};
+
+template<>
+struct StackBuffer<0> {
+    NativeWord* allocate(size_t) { VERIFY_NOT_REACHED(); }
+};
+
+inline StackBuffer<0> g_null_stack_buffer;
+
+template<typename Word, bool is_signed_>
+struct StorageSpan : Span<Word> {
+    using Span<Word>::Span;
+
+    constexpr static bool is_signed = is_signed_;
+
+    explicit constexpr StorageSpan(Span<Word> span)
+        : Span<Word>(span)
+    {
+    }
+
+    constexpr bool is_negative() const
+    {
+        return is_signed && this->last() >> (word_size - 1);
+    }
+};
+
+using UnsignedStorageSpan = StorageSpan<NativeWord, false>;
+using UnsignedStorageReadonlySpan = StorageSpan<NativeWord const, false>;
+
+// Sometimes we want to know the exact maximum amount of the bits required to represent the number.
+// However, the bit size only acts as a hint for wide multiply operations. For all other purposes,
+// `bit_size`-sized and `ceil(bit_size / word_size) * word_size`-sized `StaticStorage`s will act the
+// same.
+template<bool is_signed_, size_t bit_size>
+requires(bit_size <= max_big_int_length * word_size) struct StaticStorage {
+    constexpr static size_t static_size = (bit_size + word_size - 1) / word_size;
+    constexpr static bool is_signed = is_signed_;
+
+    // We store integers in little-endian regardless of the host endianness. We use two's complement
+    // representation of negative numbers and do not bother at all if `bit_size % word_size != 0`
+    // (i. e. do not properly handle overflows).
+    NativeWord m_data[static_size];
+
+    constexpr bool is_negative() const
+    {
+        return is_signed_ && m_data[static_size - 1] >> (word_size - 1);
+    }
+
+    constexpr static size_t size()
+    {
+        return static_size;
+    }
+
+    constexpr NativeWord operator[](size_t i) const
+    {
+        return m_data[i];
+    }
+
+    constexpr NativeWord& operator[](size_t i)
+    {
+        return m_data[i];
+    }
+
+    constexpr NativeWord const* data() const
+    {
+        return m_data;
+    }
+
+    constexpr NativeWord* data()
+    {
+        return m_data;
+    }
+};
+
+// There is no reason to ban u128{0} + 1 (although the second argument type is signed, the value is
+// known at the compile time to be non-negative). In order to do so, we provide overloads in
+// UFixedBigInt which take IntegerWrapper as an argument.
+struct IntegerWrapper {
+    StaticStorage<false, sizeof(int) * 8> m_data;
+
+    consteval IntegerWrapper(int value)
+    {
+        if (value < 0)
+            (new int)&&"Requested implicit conversion of an integer to the unsigned one will underflow.";
```

Comment:
I'm not sure if there exists any neat syntax for conditionally failing consteval function (`static_assert`, obviously, won't work here).
I got general idea how to do what I want from this SO question https://stackoverflow.com/questions/67320438/how-to-fail-a-consteval-function but throwing something will always CE because of `-fno-exceptions`, so I figured out we can just allocate something we would never deallocate.

Reply:
When maintainers approve, I won't complain

Reply:
I think calling some cleverly named non-constexpr function would be a more understandable way to do it. I believe that's what standard library implementations are doing today.

Reply:
Though I for sure want to hear @alimpfard 's thoughts on this revival of #16126 to see if the changes here and the explaination in <https://github.com/SerenityOS/serenity/pull/16126#issuecomment-1321796304> make sense to them before reviewing the patch for style.

Reply:
Organizational question: should I ping Ali on discord then?

Reply:
It's around midnight where he's at right now, so if he doesn't take a look in response to this ping (or one I already gave him separately on discord) in a day or so then a post in #code-review with a ping would be a good idea yeah.

Reply:
The ideas are pretty sound imo, I haven't actually looked at the implementation, but the stuff in https://github.com/SerenityOS/serenity/pull/16126#issuecomment-1321796304 sounds good.
I think I've mentioned before that the mixing of fixed and dynamic bigints sounds odd, but it makes sense to me that the algorithms would be generic and shared between them.

Reply:
Oh and also, calling a non-constexpr function in case you want to error in a consteval function _is_ how we do it in other parts of the system, so please prefer that.

Reply:
Yeap, sure